### PR TITLE
Create PlayerHighestRanks database entity

### LINKED ISSUE #372: Create `PlayerHighestRanks` database entity
This entity will store information about a Player's highest achieved global and country rank. Each `Player` can own a maximum of one `PlayerHighestRanks` per Ruleset (similarly to `PlayerRating`).

- [x] `PlayerHighestRanks` entity created
- [x] Navigation added to existing `Player` entity
- [x] Relevant database migration created and tested

The entity should be structured as such:
```cs
public class PlayerHighestRanks : IUpdateableEntity
{
  public Ruleset Ruleset { get; set; }

  public int GlobalRank { get; set; }
  public DateTime GlobalRankDate { get; set; }
  public int CountryRank { get; set; }
  public DateTime CountryRankDate { get; set; }

  // FK, required
  public int PlayerId { get; set; }
  // Backwards navigation
  public Player Player { get; set; }
}
```

The entity should:
- Have a table named `player_highest_ranks`
- Have a one to many relationship with `Player` (`Player` -> `PlayerHighestRanks`)
  - Unique compound index of { `PlayerId`, `Ruleset` } (Max one per ruleset per player)
- Cascade delete on `Player` deletion

### PR DESCRIPTION
**Pull request by alixdotsh**
Closes #372.

I believe I got it for the most part, but I probably will need to do some hefty changes. Please be blunt on the reviews.

### CODE REVIEW

**Comment by alixdotsh** [comment]
so glad this migration worked :sob

**Comment by hburn7** [comment]
Confirmed the generated script creates a `player_highest_ranks` table with the right fields, except the `default_ruleset` table should be named `ruleset`.

File: Database/Entities/PlayerHighestRanks.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+public class PlayerHighestRanks : UpdateableEntityBase
+{
+    /// <summary>
+    /// <see cref="Enums.Ruleset"/> as set on the <see cref="Player"/>'s osu! profile
+    /// </summary>
+    [Column("default_ruleset")]
+    public Ruleset Ruleset { get; set; }
+
+    /// <summary>
+    /// Global rank at the current time
```
Comment (hburn7) [inline Database/Entities/PlayerHighestRanks.cs:22]:
```suggestion
    /// The highest global rank for this user in the given ruleset
```

File: Database/Entities/PlayerHighestRanks.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    /// </summary>
+    [Column("global_rank")]
+    public int GlobalRank { get; set; }
+
+    /// <summary>
+    /// Date the global rank was achieved
```
Comment (hburn7) [inline Database/Entities/PlayerHighestRanks.cs:28]:
```suggestion
    /// Date this player's peak global rank was achieved in the given ruleset
```

File: Database/Entities/PlayerHighestRanks.cs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    /// </summary>
+    [Column("global_rank_date")]
+    public DateTime GlobalRankDate { get; set; }
+
+    /// <summary>
+    /// Country rank at the current time
```
Comment (hburn7) [inline Database/Entities/PlayerHighestRanks.cs:34]:
```suggestion
    /// The highest country rank for this user in the given ruleset
```

File: Database/Entities/PlayerHighestRanks.cs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    /// </summary>
+    [Column("country_rank")]
+    public int CountryRank { get; set; }
+
+    /// <summary>
+    /// Date the current country rank was achieved
```
Comment (hburn7) [inline Database/Entities/PlayerHighestRanks.cs:40]:
```suggestion
    /// Date this player's peak country rank was achieved in the given ruleset
```

File: Database/Entities/PlayerHighestRanks.cs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    /// </summary>
+    [Column("country_rank_date")]
+    public DateTime CountryRankDate { get; set; }
+
+    // FK, required
```
Comment (hburn7) [inline Database/Entities/PlayerHighestRanks.cs:45]: Add a summary tag here with `The id of the player this record belongs to`

**Comment by alixdotsh** [comment]
lemme fix that rq

File: Database/Entities/PlayerHighestRanks.cs [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    [Column("player_id")]
+    public int PlayerId { get; set; }
+
+    // Backwards navigation
+    public Player Player { get; set; } = null!;
```
Comment (hburn7) [inline Database/Entities/PlayerHighestRanks.cs:50]: Add a summary tag with `The <see cref="Player"/>` - @myssto idk if it needs to be more specific

**Review by hburn7 — CHANGES_REQUESTED** [review]

**Review by hburn7 — APPROVED** [review]

File: Database/Entities/PlayerHighestRanks.cs [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+public class PlayerHighestRanks : UpdateableEntityBase
+{
+    /// <summary>
+    /// The ruleset in which this highest rank record occurred in
+    /// </summary>
+    [Column("ruleset")]
+    public Ruleset Ruleset { get; set; }
+
+    /// <summary>
+    /// The highest global rank for this user in the given ruleset
```
Comment (myssto) [inline Database/Entities/PlayerHighestRanks.cs:22]: You can probably remove 'in the given ruleset' from all docstrings here. Its implied by the docstring on the class itself.

**Review by myssto — CHANGES_REQUESTED** [review]
Lets squash the migrations on this one. This shouldn't be done in 3 separate ones. If you undo them all and recreate a new migration with the current code it should be fine.

File: Database/Migrations/20240918214600_PlayerHighestRanks.cs
Comment (hburn7) [inline Database/Migrations/20240918214600_PlayerHighestRanks.cs:1]: This is an empty migration bruv

**Review by hburn7 — CHANGES_REQUESTED** [review]

File: Database/Migrations/20240918214600_PlayerHighestRanks.cs
Comment (alixdotsh) [reply Database/Migrations/20240918214600_PlayerHighestRanks.cs:1]: kill

Comment (hburn7) [reply Database/Migrations/20240918214600_PlayerHighestRanks.cs:1]: Fixed

**Comment by hburn7** [comment]
Have tested the current branch and can apply migrations successfully.

**Review by hburn7 — APPROVED** [review]

File: Database/Entities/PlayerHighestRanks.cs [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    public int GlobalRank { get; set; }
+
+    /// <summary>
+    /// Date this player's peak global rank was achieved
+    /// </summary>
+    [Column("global_rank_date")]
+    public DateTime GlobalRankDate { get; set; }
+
+    /// <summary>
+    /// The highest country rank for this user
```
Comment (myssto) [inline Database/Entities/PlayerHighestRanks.cs:34]:
```suggestion
    /// Highest country rank achieved
```

File: Database/Entities/PlayerHighestRanks.cs [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    /// </summary>
+    [Column("country_rank")]
+    public int CountryRank { get; set; }
+
+    /// <summary>
+    /// Date this player's peak country rank was achieved
```
Comment (myssto) [inline Database/Entities/PlayerHighestRanks.cs:40]:
```suggestion
    /// Timestamp of when the <see cref="CountryRank"/> was achieved
```

File: Database/Entities/PlayerHighestRanks.cs [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+    public DateTime CountryRankDate { get; set; }
+
+    /// <summary>
+    /// The id of the player this record belongs to
+    /// </summary>
+    [Column("player_id")]
+    public int PlayerId { get; set; }
+
+    /// <summary>
+    /// The <see cref="Player"/>
```
Comment (myssto) [inline Database/Entities/PlayerHighestRanks.cs:52]:
```suggestion
    /// The <see cref="Player"/> this record belongs to
```

File: Database/Entities/PlayerHighestRanks.cs [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    public int CountryRank { get; set; }
+
+    /// <summary>
+    /// Date this player's peak country rank was achieved
+    /// </summary>
+    [Column("country_rank_date")]
+    public DateTime CountryRankDate { get; set; }
+
+    /// <summary>
+    /// The id of the player this record belongs to
```
Comment (myssto) [inline Database/Entities/PlayerHighestRanks.cs:46]:
```suggestion
    /// The id of the <see cref="Player"/> this record belongs to
```

**Review by myssto — CHANGES_REQUESTED** [review]

File: Database/Entities/PlayerHighestRanks.cs [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+/// <summary>
+/// Represents the highest ranks of a player for a given ruleset
+/// </summary>
+[Table("player_highest_ranks")]
+[SuppressMessage("ReSharper", "CollectionNeverUpdated.Global")]
+[SuppressMessage("ReSharper", "PropertyCanBeMadeInitOnly.Global")]
+public class PlayerHighestRanks : UpdateableEntityBase
+{
+    /// <summary>
+    /// The ruleset in which this highest rank record occurred in
```
[the remark this answers is not in the document]
Comment (hburn7) [reply Database/Entities/PlayerHighestRanks.cs:16]: Would remove "in" at the end to not end sentence with preposition.

**Review by myssto — APPROVED** [review]

**Pull request merged**